crates/hulk/src/nao/double_buffered_reader.rs

```diff
@@ -0,0 +1,411 @@
+use std::{
+    io::{self, ErrorKind, Read},
+    mem::size_of,
+    os::unix::{io::AsRawFd, prelude::RawFd},
+};
+
+use super::hula::StateStorage;
+
+pub struct DoubleBufferedReader {
+    buffers: [[u8; size_of::<StateStorage>()]; 2],
+    active_buffer_index: usize,
+    incomplete_offset: usize,
+}
+
+impl Default for DoubleBufferedReader {
+    fn default() -> Self {
+        Self {
+            buffers: [[0; size_of::<StateStorage>()]; 2],
+            active_buffer_index: Default::default(),
+            incomplete_offset: Default::default(),
+        }
+    }
+}
+
+impl DoubleBufferedReader {
+    fn swap_buffers(&mut self) {
+        self.incomplete_offset = 0;
+        self.active_buffer_index = (self.active_buffer_index + 1) % 2;
+    }
+
+    pub fn read(
+        &mut self,
+        reader: &mut (impl AsRawFd + Read),
+        mut poll_reader: impl FnMut(RawFd) -> io::Result<()>,
+    ) -> io::Result<()> {
+        let mut buffers_swapped = false;
+        loop {
+            match reader.read(&mut self.buffers[self.active_buffer_index][self.incomplete_offset..])
+            {
+                Ok(number_of_read_bytes) => {
+                    self.incomplete_offset += number_of_read_bytes;
+                    assert!(self.incomplete_offset <= self.buffers[self.active_buffer_index].len());
+                    if self.incomplete_offset == self.buffers[self.active_buffer_index].len() {
```

Comment:
```suggestion
                    let is_message_complete (?) = self.incomplete_offset == self.buffers[self.active_buffer_index].len();
                    if  is_message_complete {
```
not sure about the naming, giving this comparison a name leads to better documented behavior

Reply:
Is extracted as `is_active_buffer_complete` to align with the `is_at_least_one_buffer_complete` variable.